### PR TITLE
Tests for extract_rings() and handling of `Unlisted` internally

### LINKED ISSUE #36: Some rings are not split correctly - write tests!
Whilst tinkering and working on #24 (see solution in #35) I noticed that some of the ring combinations were not being
split correctly.

- `U` wrong code causing an error, removed as part of #35
- `gn` only get BTO assigned, e.g. `gnBL` puts BTO as Top Ring on right leg, which is correct but should have `Split
  green & black (gn)` Top Left and `Light blue (B)` bottom right
- `ngDL` same problem as above
- `ryFR` ditto

There may well be more, we will have to carefully check all of these. There are some +700 so writing formal tests for this
would be the way forward.

(Ashamed to say I've written virtually no tests, I did start on some but have almost zero experience of writing tests in
R and felt I had to focus on getting things working instead, I should have known better!).

### PR DESCRIPTION
**Pull request by ns-rse**
Closes #36 

**NB** This is a stacked pull request and should be merged _after_ #45 

Noted in #36 was that some of the split rings (those with two colours) were not being correctly handled. Now that we have a package structure in place (see #45 on which this branch was forked from) we can use the formal [testthat framework](https://testthat.r-lib.org/) to run the tests, some of which had already been written but were not correctly structured.

The problematic ring combinations noted in #36 are now explicitly tested and are correctly handled via pararmeterisation[^1]. We also now allow `"Unlisted"` as an option and it is tested to return an empty set of values for `rings` for string items (although the `rings$pit` is returned as `NA` as it is normally boolean). This will allow us to now add `shiny::textInput()` fields for entering individual rings.

- Corrects `UL` > `UUL` as described in #36 comments.
- Correct layout/structure of `testthat` unittests.
- Parameterise tests of `lottie::extracted_rings()`.
- We will be allowing users to select "Unlisted" from the ring options, we therefore add this into the dataframe defined in `lookups.R`.
- `lottie::extract_rings()` now handles `Unlisted` option, returning emptry strings (`""`) or `NA`.
- Expand test suite to check the failures identified in #36
- Adds tests for...
  - `lottie::tidy_columns()`
  - `lottie::remove_none_column()`
  - `lottie::deduplicate_flock()`

[^1]: No change to `lottie::expected_rings()` was required here though, it looks like it may well have already been fixed at some point.

### CODE REVIEW

File: R/utils.R [hunk 1]
```diff
@@ -193,14 +142,37 @@ remove_none_column <- function(df) {
...
+#'  |`pit_pos` | str     | The position of the "pit" ring.                          |
+#'  |`bto`     | str     | The leg the BTO ring is on (inferred).                   |
+#'  |`first`   | str     | The first ring from the full ring code.                  |
+#'  |`second`  | str     | The second ring from the full ring code.                 |
+#'
+#' @export
 extract_rings <- function(code, valid_codes, known_rings) {
     ## Validate that the supplied code is valid
     if (!(code %in% known_rings)) {
         print(print0("WARNING!!! The provided combination (", code, ") is not in known_rings."))
```
Comment (tdjames1) [inline R/utils.R]: @ns-rse A tpyo here stopped tests running for me
```suggestion
        print(paste0("WARNING!!! The provided combination (", code, ") is not in known_rings."))
```

**Comment by ns-rse** [comment]
Rebased post merge of #45 and ready for review.

File: tests/testthat/test-utils.R [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+patrick::with_parameters_test_that("Splitting of all codes is correct:",
+  {
+    testthat::expect_identical(extracted_rings, expected_rings)
+  },
+  extracted_rings = c(
+    lottie::extract_rings(code = "bmrdR", valid_codes = valid_codes, known_rings = known_rings),
+    lottie::extract_rings(code = "SdnrL", valid_codes = valid_codes, known_rings = known_rings),
+    lottie::extract_rings(code = "PonR", valid_codes = valid_codes, known_rings = known_rings ),
+    lottie::extract_rings(code = "WY*L", valid_codes = valid_codes, known_rings = known_rings),
+    lottie::extract_rings(code = "WY*L", valid_codes = valid_codes, known_rings = known_rings),
```
Comment (tdjames1) [inline tests/testthat/test-utils.R]: Duplicate cases?

File: tests/testthat/test-utils.R [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    lottie::extract_rings(code = "ryOR", valid_codes = valid_codes, known_rings = known_rings),
+    lottie::extract_rings( code = "SgNL", valid_codes = valid_codes, known_rings = known_rings),
+    lottie::extract_rings( code = "BDR", valid_codes = valid_codes, known_rings = known_rings),
+    lottie::extract_rings( code = "BDL", valid_codes = valid_codes, known_rings = known_rings),
+    lottie::extract_rings( code = "gnBL", valid_codes = valid_codes, known_rings = known_rings),
+    lottie::extract_rings( code = "ngDL", valid_codes = valid_codes, known_rings = known_rings),
+    lottie::extract_rings( code = "ryFR", valid_codes = valid_codes, known_rings = known_rings),
+    lottie::extract_rings( code = "Unlisted", valid_codes = valid_codes, known_rings = known_rings),
+    lottie::extract_rings( code = "None", valid_codes = valid_codes, known_rings = known_rings)
```
Comment (tdjames1) [inline tests/testthat/test-utils.R]: [nit]
```suggestion
    lottie::extract_rings(code = "SgNL", valid_codes = valid_codes, known_rings = known_rings),
    lottie::extract_rings(code = "BDR", valid_codes = valid_codes, known_rings = known_rings),
    lottie::extract_rings(code = "BDL", valid_codes = valid_codes, known_rings = known_rings),
    lottie::extract_rings(code = "gnBL", valid_codes = valid_codes, known_rings = known_rings),
    lottie::extract_rings(code = "ngDL", valid_codes = valid_codes, known_rings = known_rings),
    lottie::extract_rings(code = "ryFR", valid_codes = valid_codes, known_rings = known_rings),
    lottie::extract_rings(code = "Unlisted", valid_codes = valid_codes, known_rings = known_rings),
    lottie::extract_rings(code = "None", valid_codes = valid_codes, known_rings = known_rings)
```

File: tests/testthat/test-utils.R [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+valid_codes <- c(
+    "BDL", "BDR", ## Length 3
+    "SgNL", "ryOR", ## Length 4, first ring is two characters
+    "WY*L", "PonR", ## Length 4, first ring is one character
+    "SdnrL", "bmrdR", ## Length 5
+    "gnBL", "ngDL", "ryFR", ## Split rings
+    "Unlisted"
+)
+
+patrick::with_parameters_test_that("Splitting of all codes is correct:",
```
Comment (tdjames1) [inline tests/testthat/test-utils.R]:
```suggestion
patrick::with_parameters_test_that("Splitting of known codes is correct:",
```

File: tests/testthat/test-utils.R [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    testthat::expect_false("none" %in% tidy_columns)
+})
+
+
+testthat::test_that("Deduplication of ringed birds only", {
+    deduplicated <- lottie::deduplicate_flock(df = data.frame(flock_id = c(1, 1, 1, 1), ringed = c(FALSE, FALSE, TRUE, TRUE)))
+    testthat::expect_equal(deduplicated, data.frame(flock_id = c(1, 1, 1), ringed = c(FALSE, FALSE, TRUE)))
+})
+
+patrick::with_parameters_test_that("Deduplication of ringed birds only:",
```
Comment (tdjames1) [inline tests/testthat/test-utils.R:88]: Does this supercede the test at [lines 87-89](https://github.com/Morinay-Lab/lotties/pull/48/changes#diff-f3738eec05b13578eefad27b708d2579de254f58f3b9a8fd88b7c5086399e75fR87-R89)?

File: tests/testthat/test-utils.R [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+        "Sw", "Sy", "bm", "dg", "do", "dr", "dy", "gd", "gn", "mb", "ng", "no", "nr", "od", "on", "rd", "rn", "ry",
+        "yd", "yr", "B*", "G*", "N*", "P*", "R*", "W*", "Y*")
+valid_codes <- c(
+    "BDL", "BDR", ## Length 3
+    "SgNL", "ryOR", ## Length 4, first ring is two characters
+    "WY*L", "PonR", ## Length 4, first ring is one character
+    "SdnrL", "bmrdR", ## Length 5
+    "gnBL", "ngDL", "ryFR", ## Split rings
+    "Unlisted"
+)
```
Comment (tdjames1) [inline tests/testthat/test-utils.R]: In `server.R`, `extract_rings()` is called with arguments:

```
lottie::extract_rings(
            code = input$composition_colour_ring,
            valid_codes = rings_df$code,
            known_rings = colour_ring_df$code)
```

where `rings_df` is the list of individual colour rings and `colour_ring_df` is the list of known colour ring combinations. So I think that these variable names are the wrong way round.

**Review by tdjames1 — CHANGES_REQUESTED** [review]
Hi @ns-rse I'm requesting changes because I think there is a bug in the test script. Please see the comment on the input parameters for the tests for `extract_rings()` (`test-utils.R`, lines 5-16).

Correcting these variable names uncovers a bug in `extract_rings()` (again due to confusion between `valid_codes` and `known_rings`)  which I was able to resolve with the following patch:
```
diff --git i/R/utils.R w/R/utils.R
index 7486da9..e634c90 100644
--- i/R/utils.R
+++ w/R/utils.R
@@ -235,12 +235,12 @@ extract_rings <- function(code, valid_codes, known_rings) {
     } else if (code_length == 4) {
         ## We check to see if the first two characters are in the subset of rings that are two characters in length, if so we
         ## use the first two characters as the top and the third is the bottom
-        known_rings2 <- known_rings[stringr::str_length(known_rings) == 2]
+        valid_codes2 <- valid_codes[stringr::str_length(valid_codes) == 2]
         if (code == "None") {
             rings$leg <- ""
             rings$first <- ""
             rings$second <- ""
-        } else if (stringr::str_sub(code, 1, 2) %in% known_rings2) {
+        } else if (stringr::str_sub(code, 1, 2) %in% valid_codes2) {
             rings$first <- stringr::str_sub(code, 1, 2)
             rings$second <- stringr::str_sub(code, 3, 3)
         ## If not then the first character is the top and the second and third are the bottom.
```

It would probably also be good to adjust the `extract_rings()` function documentation to clarify what `valid_codes` and `known_rings` represent.

File: tests/testthat/test-utils.R (hunk 5 above)
Comment (ns-rse) [reply tests/testthat/test-utils.R:88]: Yes it does, I wrote the single case test first and meant to remove it after parameterising the test and clearly forgot to remove it. Have done so no.

File: tests/testthat/test-utils.R (hunk 2 above)
Comment (ns-rse) [reply tests/testthat/test-utils.R]: Oops, meant to switch order and test when PIT ring is first, thanks for spotting, updated.

File: tests/testthat/test-utils.R (hunk 6 above)
Comment (ns-rse) [reply tests/testthat/test-utils.R]: :facepalm: sorry and thank you for picking this up.

**Comment by ns-rse** [comment]
Thanks for spotting all these errors and improvements @tdjames1. I've (hopefully!) corrected them and clarified nomenclature.

**Review by tdjames1 — CHANGES_REQUESTED** [review]
Tests are now passing but there are some occurrences of `rings_df` in `ui.R` and `server.R` which need to be fixed.

**Comment by ns-rse** [comment]
> Tests are now passing but there are some occurrences of `rings_df` in `ui.R` and `server.R` which need to be fixed.

Amended. I shouldn't have rushed to do this on Friday whilst working on my other project for the day, was keen to get things pushed through and progress. Reminder to myself that context switching is ultimately detrimental.

File: R/utils.R [hunk 7]
```diff
@@ -160,8 +140,11 @@ remove_none_column <- function(df) {
 #' Rings} for further details on the rules used in the coding system.
 #'
 #' @param code str The code from which rings and leg are to be extracted.
-#' @param valid_codes list List of valid codes, if `code` is not in `valid_codes` it is not extracted.
-#' @param known_rings list List of rings used.
+#' @param valid_ring_combinations list A list of valid ring combinations of the form "[ring1][ring2][leg]", e.g. `Y*WL`, `SdnrL`,
+#' `ryFR`, `BDL`. These will typically be taken from the ``
```
Comment (tdjames1) [inline R/utils.R]: Something missing here?

File: R/utils.R [hunk 8]
```diff
@@ -176,14 +159,23 @@ remove_none_column <- function(df) {
 #'  |`second`  | str     | The second ring from the full ring code.                 |
 #'
 #' @export
-extract_rings <- function(code, valid_codes, known_rings) {
+extract_rings <- function(code, valid_ring_combinations, valid_rings) {
```
Comment (tdjames1) [inline R/utils.R:165]: Having modified the arguments, need to update the associated documentation.

File: tests/testthat/test-utils.R [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+    list(" code" = "Y*WL"," leg" = "L"," pit" = TRUE," bto" = "R"," first" = "Y*"," second" = "W"," pit_pos" = "first"),
+    list("code" = "ryOR", "leg" = "R", "pit" = FALSE, "bto" = "L", "first" = "ry", "second" = "O"),
+    list("code" = "SgNL", "leg" = "L", "pit" = FALSE, "bto" = "R", "first" = "Sg", "second" = "N" ),
+    list("code" = "BDR", "leg" = "R", "pit" = FALSE, "bto" = "L", "first" = "B", "second" = "D"),
+    list( "code" = "BDL", "leg" = "L", "pit" = FALSE, "bto" = "R", "first" = "B", "second" = "D"),
+    list( "code" = "gnBL", "leg" = "L", "pit" = FALSE, "bto" = "R", "first" = "gn", "second" = "B"),
+    list( "code" = "ngDL", "leg" = "L", "pit" = FALSE, "bto" = "R", "first" = "ng", "second" = "D"),
+    list( "code" = "ryFR", "leg" = "R", "pit" = FALSE, "bto" = "L", "first" = "ry", "second" = "F"),
+    list( "code" = "Unlisted", "leg" = "", "pit" = NA, "bto" = "", "first" = "", "second" = ""),
+    list( "code" = "None", "leg" = "", "pit" = NA, "bto" = "None", "first" = "", "second" = "")
```
Comment (tdjames1) [inline tests/testthat/test-utils.R]: Test fails due to expected value `NA` for `pit`. Value gets set to `FALSE` at line 188 in `utils.R`.

**Review by tdjames1 — CHANGES_REQUESTED** [review]
Hi @ns-rse sorry to be a pain but there is now a failing test as noted in my inline comments. 

Also, I think you need to rebuild `.Rd` to match changes to documentation.

**Comment by ns-rse** [comment]
Sorry, I'm clearly overly reliant on pre-commit hooks to run tests for me (and again am task switching having been working on trying to handle the reactive logic for updating ring fields). Will fix these now (run the tests and check the UI too).

File: R/utils.R (hunk 8 above)
Comment (ns-rse) [reply R/utils.R:165]: I thought I had done that, the above docstrings reflect these new variable names and the descriptions have been expanded from their previous short forms.

Comment (ns-rse) [reply R/utils.R:165]: After some head scratching (and reading the full message rather than inline ones) I realised you meant the `man/*.Rd` files now and have now updated that.

I shouldn't make that mistake again as I'm now using the R pre-commit `roxygenize` [hook](https://lorenzwalthert.github.io/precommit/articles/available-hooks.html) so that should automatically update on commits.

Comment (tdjames1) [reply R/utils.R:165]: Sorry for being unclear!

**Review by tdjames1 — APPROVED** [review]
Thanks @ns-rse all good now!

File: R/utils.R (hunk 8 above)
Comment (ns-rse) [reply R/utils.R:165]: No you were clear, it was me who missed it, I dived straight-in reading the individual comments rather than the overall ones.

**Comment by ns-rse** [comment]
> Thanks @ns-rse all good now!

Thanks for your endless patience @tdjames1
